// react-express-router/dist/elements/Express.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const types_1 = require("../types");
const utils_1 = require("../utils");
const Node_1 = __importDefault(require("./Node"));
const errorHandler_1 = require("../validation/errorHandler");
class ExpressInstance extends Node_1.default {
    constructor(type, props) {
        super(type, props);
        this.props = Object.assign(Object.assign({}, props), { path: props && props.path ? props && props.path : utils_1.DEFAULT_PATH });
    }
    render() {
        const nodes = this.childs.map(c => c.render());
        const sortedNodes = nodes.sort(sortErrorHandler);
        errorHandler_1.twiceErrorHandler(sortedNodes);
        return {
            type: types_1.Elements.Express,
            path: this.props.path,
            routes: sortedNodes,
        };
    }
}
exports.default = ExpressInstance;
function sortErrorHandler(a, b) {
    return a.type === types_1.Elements.ErrorHandler ? 1 : 0;
}
