// react-express-router/dist/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Elements;
(function (Elements) {
    Elements["Route"] = "ROUTE";
    Elements["Middleware"] = "MIDDLEWARE";
    Elements["ParamMiddleware"] = "PARAMMIDDLEWARE";
    Elements["Express"] = "EXPRESS";
    Elements["Root"] = "ROOT";
    Elements["ErrorHandler"] = "ERRORHANDLER";
})(Elements = exports.Elements || (exports.Elements = {}));
var Methods;
(function (Methods) {
    Methods["CHECKOUT"] = "checkout";
    Methods["COPY"] = "copy";
    Methods["DELETE"] = "delete";
    Methods["GET"] = "get";
    Methods["HEAD"] = "head";
    Methods["LOCK"] = "lock";
    Methods["MERGE"] = "merge";
    Methods["MKACTIVITY"] = "mkactivity";
    Methods["MKCOL"] = "mkcol";
    Methods["MOVE"] = "move";
    Methods["MSEARCH"] = "m-search";
    Methods["NOTIFY"] = "notify";
    Methods["OPTIONS"] = "options";
    Methods["PATCH"] = "patch";
    Methods["POST"] = "post";
    Methods["PURGE"] = "purge";
    Methods["PUT"] = "put";
    Methods["REPORT"] = "report";
    Methods["SEARCH"] = "search";
    Methods["SUBSCRIBE"] = "subscribe";
    Methods["TRACE"] = "trace";
    Methods["UNLOCK"] = "unlock";
    Methods["UNSUBSCRIE"] = "unsubscribe";
    Methods["ALL"] = "all";
})(Methods = exports.Methods || (exports.Methods = {}));
